### ISSUE TITLE
Macro reification of abstract/final class/interface

### ISSUE THREAD

**Issue by ibilon** [HaxeFoundation/haxe#10060]
Tried doing an abstract or final class through macro reification, but using modifiers doesn't work:
```haxe
macro abstract class A {};
macro final class B {};
macro abstract interface C {};
macro final interface D {};
```
but got the error `Expected expression` for `abstract` and `Missing ;` for `final`.

As a workaround I can fix the `Typedefinition.kind` after the fact.
```haxe
var cls = macro class A {
  abstract function test():Void;
}
cls.kind = TDClass(null, [], false, false, true);
```
but it's a bit hacky.

Might be out of scope, but I get the same error if I try to make an abstact, enum or enum abstract:
```haxe
macro abstract A(Int) {};
macro enum abstract B(Int) {};
macro enum C {};
```

**Comment by back2dos** [comment]
Related: `macro @meta class {}` also doesn't parse.

**Comment by PXshadow** [comment]
To add to this, private declaring of typedefs and classes does not work.
```haxe
var expr = macro private class Test {};
```
There is also no Access method in TypeDefinition, except when using, kind:  ``TDField(kind:FieldType, access:Array<Access>)``

**Comment by Simn** [comment]
Yeah this is tricky... We're crossing the boundaries between expression and type parsing here. To make matters worse, I don't want to completely block off the possibility of having field reification at some point.

Let's make a table:

what | type? | field? | expr? | idea?
--- | --- | --- | --- | ---
`abstract` | yes | yes | no | field is rare, use type
`enum` | yes | no | no | no contest, use type
`@meta` | yes | yes | yes | recurse & check again
`final` | yes | yes | yes | hmmm...
`private` | yes | yes | no | not sure
`public` | no | yes | no | use field

I was thinking that it would be nice if both `public` and `private` entered field-mode. This would make their usage mandatory, but that seems fair to distinguish reification modes. This would then also free up `abstract` for type mode, because if we want a field we can just be explicit about it.

However, then we can't easily reify `private` types. Maybe that's a fair limitation because it never worked anyway and you can always manually add the private flag if you want it.

The real problem here is `final` because that can be just about anything. If we go with the `public`/`private` for types, we can at least ignore fields, but we still have to check if we want a type or an expression.
